github_issue_summarization/distributed/train.py

```diff
@@ -14,164 +15,208 @@
+
+    # Save the processed data
+    np.save(data_dir + 'train_title_vecs.npy', train_title_vecs)
+    np.save(data_dir + 'train_body_vecs.npy', train_body_vecs)
+  else:
+    time.sleep(120)
+
+  # TODO(jlewi): Why do we need to block waiting for the file?
+  # I think this is because only the master produces the npy
+  # files so the other workers need to wait for the files to arrive.
+  # It might be better to make preprocessing a separate job.
+  while True:
+    if os.path.isfile(data_dir + 'train_body_vecs.npy'):
+      break
+    print("Waiting for dataset")
+    time.sleep(2)
+  encoder_input_data, doc_length = load_encoder_inputs(data_dir + 'train_body_vecs.npy')
+  decoder_input_data, decoder_target_data = load_decoder_inputs(data_dir + 'train_title_vecs.npy')
 
-while True:
-  if os.path.isfile(data_dir + 'train_body_vecs.npy'):
-    break
-  print("Waiting for dataset")
-  time.sleep(2)
-encoder_input_data, doc_length = load_encoder_inputs(data_dir + 'train_body_vecs.npy')
-decoder_input_data, decoder_target_data = load_decoder_inputs(data_dir + 'train_title_vecs.npy')
+  num_encoder_tokens, body_pp = load_text_processor(data_dir + 'body_pp.dpkl')
+  num_decoder_tokens, title_pp = load_text_processor(data_dir + 'title_pp.dpkl')
 
-num_encoder_tokens, body_pp = load_text_processor(data_dir + 'body_pp.dpkl')
-num_decoder_tokens, title_pp = load_text_processor(data_dir + 'title_pp.dpkl')
+  #arbitrarly set latent dimension for embedding and hidden units
+  latent_dim = 300
 
-#arbitrarly set latent dimension for embedding and hidden units
-latent_dim = 300
+  ##### Define Model Architecture ######
 
-##### Define Model Architecture ######
+  ########################
+  #### Encoder Model ####
+  encoder_inputs = tf.keras.layers.Input(shape=(doc_length,), name='Encoder-Input')
 
-########################
-#### Encoder Model ####
-encoder_inputs = tf.keras.layers.Input(shape=(doc_length,), name='Encoder-Input')
+  # Word embeding for encoder (ex: Issue Body)
+  x = tf.keras.layers.Embedding(
+        num_encoder_tokens, latent_dim, name='Body-Word-Embedding', mask_zero=False)(encoder_inputs)
+  x = tf.keras.layers.BatchNormalization(name='Encoder-Batchnorm-1')(x)
 
-# Word embeding for encoder (ex: Issue Body)
-x = tf.keras.layers.Embedding(
-      num_encoder_tokens, latent_dim, name='Body-Word-Embedding', mask_zero=False)(encoder_inputs)
-x = tf.keras.layers.BatchNormalization(name='Encoder-Batchnorm-1')(x)
+  # Intermediate GRU layer (optional)
+  #x = GRU(latent_dim, name='Encoder-Intermediate-GRU', return_sequences=True)(x)
+  #x = BatchNormalization(name='Encoder-Batchnorm-2')(x)
 
-# Intermediate GRU layer (optional)
-#x = GRU(latent_dim, name='Encoder-Intermediate-GRU', return_sequences=True)(x)
-#x = BatchNormalization(name='Encoder-Batchnorm-2')(x)
+  # We do not need the `encoder_output` just the hidden state.
+  _, state_h = tf.keras.layers.GRU(latent_dim, return_state=True, name='Encoder-Last-GRU')(x)
```

Comment:
@inc0 @hamelsmu Doesn't look like state_h is used anywhere?

Reply:
It was originally supposed to be used in line 166

`encoder_model = tf.keras.Model(inputs=encoder_inputs, outputs=state_h, name='Encoder-Model')`

Encapsulating a subset of the layers as a model object makes it easier to extract the encoder as it logically groups together those layers under the name "Encoder Model".  However, depending on the use case you are demonstrating here maybe you do not care about this?  

However, per my other comment, you will eventually need to initialize the decoder with the last hidden state of the encoder which was formerly done via the commented out `encoder_model` object.